CRM/Contribute/Form/Contribution.php

```diff
@@ -883,7 +883,7 @@ public static function formRule($fields, $files, $self) {
 
     if (!empty($fields['total_amount']) && (!empty($fields['net_amount']) || !empty($fields['fee_amount']))) {
       $sum = CRM_Utils_Rule::cleanMoney($fields['net_amount']) + CRM_Utils_Rule::cleanMoney($fields['fee_amount']);
-      if (CRM_Utils_Rule::cleanMoney($fields['total_amount']) != $sum) {
+      if (round(CRM_Utils_Rule::cleanMoney($fields['total_amount'])) != round($sum)) {
```

Comment:
@yashodha can you please allow precision to 2 i.e. ```round($sum, 2)``` as done in all other places on amount calculation.

Reply:
@yashodha Ok, I did a bit of debugging on this, basically the issue is that we're doing a string|float comparision so it doesn't like it.  The safest fix seems to be to cast the `$sum` to a string before comparing (as cleanMoney already returns a string):
`if (CRM_Utils_Rule::cleanMoney($fields['total_amount']) != (string) $sum) {`

Reply:
Hmm interesting. Thanks @mattwire for getting the exact cause (should have gone through the cleanMoney(...) code). On second thought I think we should make comparison between money (/float) values i.e.  ```if ($fields['total_amount'] != $sum)```

Like we are already doing calculation on other places like [here](https://github.com/civicrm/civicrm-core/blob/master/CRM/Contribute/Form/AdditionalPayment.php#L305)

Reply:
> Like we are already doing such calculation inside formRule, on other places like here

@monishdeb I'm not sure that copes well with currency formatted with "," as decimal place?  In fact I think that code in AdditionalPayment will fail for those locales?  I'm not an expert, but there seem to be issues comparing floats directly in PHP (according to google) as the internal data type means they can have a very small difference in precision and thus not quite be equal.  (I also don't really like doing comparison as a string, but then we either convert one to a string, or the other to a float (which I did try first but didn't seem to work).